### PR TITLE
Add unittest for SSL client API

### PR DESCRIPTION
**Pull request by andreas-bok-sociomantic**
Check that options and context are set up correctly
for encrypted connections

Blocked on #291

### CODE REVIEW

**Comment by andreas-bok-sociomantic** [comment]
Unblocked after #291 was merged

File: tests/unit/include.am [hunk 1]
```diff
@@ -148,3 +148,15 @@ tests_unit_statement_nulls_LDADD= src/libdrizzle-redux@LIBDRIZZLE_MAJOR@.la
 nodist_EXTRA_tests_unit_statement_nulls_SOURCES = dummy.cxx
 check_PROGRAMS+= tests/unit/statement_nulls
 noinst_PROGRAMS+= tests/unit/statement_nulls
+
+tests_unit_ssl_SOURCES= tests/unit/ssl.c
+tests_unit_ssl_LDADD= src/libdrizzle-redux@LIBDRIZZLE_MAJOR@.la
+nodist_EXTRA_tests_unit_ssl_SOURCES = dummy.cxx
```
Comment (ben-palmer-sociomantic) [inline tests/unit/include.am]: Extra space before `=`.

File: tests/unit/ssl.c [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+ *  tested.
+ */
+int main(int argc, char *argv[]) {
+  (void)argc;
+  (void)argv;
+  drizzle_return_t driz_ret;
+  char client_key[256];
+  char client_cert[256];
+  char server_ca[256];
+  const char * ca_path;
```
Comment (ben-palmer-sociomantic) [inline tests/unit/ssl.c]: Extra space after `*`.

File: tests/unit/ssl.c [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+                        getenv("MYSQL_USER"), getenv("MYSQL_PASSWORD"),
+                        getenv("MYSQL_SCHEMA"), 0);
+  ASSERT_NOT_NULL_(con, "Drizzle connection object creation error");
+
+  driz_ret = drizzle_connect(con);
+  ASSERT_EQ_(DRIZZLE_RETURN_OK, driz_ret, "%s(%s)", drizzle_error(con),
+             drizzle_strerror(driz_ret));
+
+  // Check if server has SSL capabilities enabled.
+  SKIP_IF_((drizzle_capabilities(con) & DRIZZLE_CAPABILITIES_SSL) == 0,
```
Comment (ben-palmer-sociomantic) [inline tests/unit/ssl.c]: Why are we doing a `SKIP_IF_` here? Wouldn't that mean this test isn't doing anything?

File: tests/unit/ssl.c [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+ *  SSL filenames
+ *   - client-key.pem
+ *   - client-cert.pem
+ *   - ca.pem
+ *
+ *  The path to the certificate and key files must be specified with the
+ *  environment variable DRIZZLE_MYSQL_CA_PATH
+ *
+ *  Actually establishing the encrypted connection the MySQL server is not
+ *  tested.
```
Comment (ben-palmer-sociomantic) [inline tests/unit/ssl.c]: This comment seems unclear to me when the connection is being created. Could you expand it a bit?

File: tests/unit/ssl.c [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    "SSL not enabled for connection to MySQL server. Skipping test");
+
+  // Check environment variable DRIZZLE_MYSQL_CA_PATH is set
+  ca_path = getenv("DRIZZLE_MYSQL_CA_PATH");
+  ASSERT_NOT_NULL_(ca_path,
+    "DRIZZLE_MYSQL_CA_PATH (path to certificate and key files) not set");
+
+  sprintf(client_key, "%s/%s", ca_path ,"client-key.pem");
+  sprintf(client_cert, "%s/%s", ca_path ,"client-cert.pem");
+  sprintf(server_ca, "%s/%s", ca_path ,"ca.pem");
```
Comment (ben-palmer-sociomantic) [inline tests/unit/ssl.c]: Spaces after `ca_path` are incorrect in the above 3 lines.

File: tests/unit/ssl.c [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+  ASSERT_EQ_(DRIZZLE_RETURN_SSL_ERROR, driz_ret, "drizzle_set_ssl(): %s(%s)",
+    drizzle_error(con), drizzle_strerror(driz_ret));
+
+  driz_ret = drizzle_set_ssl(con, client_key, client_cert, "invalid_ssl_ca",
+    ca_path, NULL);
+  ASSERT_EQ_(DRIZZLE_RETURN_SSL_ERROR, driz_ret, "drizzle_set_ssl(): %s(%s)",
+    drizzle_error(con), drizzle_strerror(driz_ret));
+
+  driz_ret = drizzle_set_ssl(con, client_key, client_cert, server_ca, ca_path,
+    cipher);
```
Comment (ben-palmer-sociomantic) [inline tests/unit/ssl.c:107]: Should there also be other unit tests for `drizzle_set_ssl` calls that don't provide all the parameters or are they all needed (I thought you only needed the certificate path)?

File: tests/unit/ssl.c [hunk 7]
```diff
@@ -0,0 +1,112 @@
+/*  vim:expandtab:shiftwidth=2:tabstop=2:smarttab:
+ *
+ *  Drizzle Client & Protocol Library
+ *
+ * Copyright (C) 2018 Sociomantic Tsunami
```
Comment (ben-palmer-sociomantic) [inline tests/unit/ssl.c]: This copyright message should just be the normal sociomantic copyright message.

**Review by ben-palmer-sociomantic — COMMENTED** [review]
Had a few comments and queries.

File: tests/unit/ssl.c (hunk 3 above)
Comment (andreas-bok-sociomantic) [reply tests/unit/ssl.c]: > Why are we doing a SKIP_IF_ here? Wouldn't that mean this test isn't doing anything?

Yes. The reason is that a user might not have set up SSL for the MySQL installation. 

In a sense it was to allow the travis ci test to succeed. But the last commit fixes support for SSL tests on travis. Will remove the SKIP_IF

File: tests/unit/ssl.c (hunk 4 above)
Comment (andreas-bok-sociomantic) [reply tests/unit/ssl.c]: Yes. It is basically that connecting using SSL to the MySQL server doesn't work without a workaround which disables non-blocking socket regardless of the options set by the client. Thus breaking the library for clients which uses non-blocking. Will rephrase it to make it more clear though

File: tests/unit/ssl.c (hunk 7 above)
Comment (andreas-bok-sociomantic) [reply tests/unit/ssl.c]: Like this? 

>   Copyright: Copyright (c) 2018 sociomantic labs GmbH?

**Comment by andreas-bok-sociomantic** [comment]
Addressed comments by @ben-palmer-sociomantic

Previous head was a0dcb07, new **HEAD** at f395c46.

<details>
<summary>Diff between heads are:</summary>

```diff

 tests/unit/include.am |  2 +-
 tests/unit/ssl.c      | 13 +++++++------
 2 files changed, 8 insertions(+), 7 deletions(-)

diff --git a/tests/unit/include.am b/tests/unit/include.am
index cde8d9b..4442387 100644
--- a/tests/unit/include.am
+++ b/tests/unit/include.am
@@ -151,7 +151,7 @@ noinst_PROGRAMS+= tests/unit/statement_nulls
 
 tests_unit_ssl_SOURCES= tests/unit/ssl.c
 tests_unit_ssl_LDADD= src/libdrizzle-redux@LIBDRIZZLE_MAJOR@.la
-nodist_EXTRA_tests_unit_ssl_SOURCES = dummy.cxx
+nodist_EXTRA_tests_unit_ssl_SOURCES= dummy.cxx
 check_PROGRAMS+= tests/unit/ssl
 noinst_PROGRAMS+= tests/unit/ssl
 
diff --git a/tests/unit/ssl.c b/tests/unit/ssl.c
index 0ba53b4..a77ed11 100644
--- a/tests/unit/ssl.c
+++ b/tests/unit/ssl.c
@@ -50,8 +50,9 @@
  *  The path to the certificate and key files must be specified with the
  *  environment variable DRIZZLE_MYSQL_CA_PATH
  *
- *  Actually establishing the encrypted connection the MySQL server is not
- *  tested.
+ *  Note that establishing the encrypted connection to the MySQL server is not
+ *  tested. This is due to a bug where openssl doesn't work with an underlying
+ *  non-blocking socket.
  */
 int main(int argc, char *argv[]) {
   (void)argc;
@@ -60,7 +61,7 @@ int main(int argc, char *argv[]) {
   char client_key[256];
   char client_cert[256];
   char server_ca[256];
-  const char * ca_path;
+  const char *ca_path;
   const char *cipher = "AES256-SHA";
   drizzle_st *con;
 
@@ -76,8 +77,8 @@ int main(int argc, char *argv[]) {
              drizzle_strerror(driz_ret));
 
   // Check if server has SSL capabilities enabled.
-  SKIP_IF_((drizzle_capabilities(con) & DRIZZLE_CAPABILITIES_SSL) == 0,
-    "SSL not enabled for connection to MySQL server. Skipping test");
+  ASSERT_NEQ_((drizzle_capabilities(con) & DRIZZLE_CAPABILITIES_SSL), 0,
+    "SSL not supported by the MySQL server");
 
   // Check environment variable DRIZZLE_MYSQL_CA_PATH is set
   ca_path = getenv("DRIZZLE_MYSQL_CA_PATH");
@@ -86,7 +87,7 @@ int main(int argc, char *argv[]) {
 
   sprintf(client_key, "%s/%s", ca_path ,"client-key.pem");
   sprintf(client_cert, "%s/%s", ca_path ,"client-cert.pem");
-  sprintf(server_ca, "%s/%s", ca_path ,"ca.pem");
+  sprintf(server_ca, "%s/%s", ca_path,"ca.pem");
 
   driz_ret = drizzle_set_ssl(con, "invalid_ssl_key", client_cert, server_ca,
     ca_path, NULL);

```
</details>

**Comment by andreas-bok-sociomantic** [comment]
Fixed copyright line

Previous head was f395c46, new **HEAD** at c0363bd.

<details>
<summary>Diff between heads are:</summary>

```diff

 tests/unit/ssl.c | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/tests/unit/ssl.c b/tests/unit/ssl.c
index a77ed11..0a0db66 100644
--- a/tests/unit/ssl.c
+++ b/tests/unit/ssl.c
@@ -2,8 +2,7 @@
  *
  *  Drizzle Client & Protocol Library
  *
- * Copyright (C) 2018 Sociomantic Tsunami
- * All rights reserved.
+ * Copyright: Copyright (c) 2018 sociomantic labs GmbH. All rights reserved
  *
  * Redistribution and use in source and binary forms, with or without
  * modification, are permitted provided that the following conditions are

```
</details>

**Comment by andreas-bok-sociomantic** [comment]
Rebased on latest v6.x.x

**Comment by ben-palmer-sociomantic** [comment]
LGTM
